File: src/server/src/main/java/com/project/application/models/exercise/ExerciseDetailsResource.java
package com.project.application.models.exercise;

import com.project.domain.workout.Exercise;
import io.micronaut.serde.annotation.Serdeable;

import java.util.List;
import java.util.UUID;

@Serdeable
public record ExerciseDetailsResource(UUID id, String name, String gifUrl, String translatedName,
                                      String description, List<MuscleGroupDetailsResource> muscleGroups,
                                      List<ExerciseResource> variations, List<EquipmentResource> equipment,
                                      List<String> mainMuscleGroupsIds, List<String> synergisticMuscleGroupsIds) {

    public static ExerciseDetailsResource from(Exercise exercise) {
        return new ExerciseDetailsResource(exercise.id, exercise.getName(),
                exercise.getGifUrl(), exercise.getTranslatedName(), exercise.getDescription(),
                exercise.getMuscleGroups().stream().map(MuscleGroupDetailsResource::from).toList(),
                exercise.getVariations().stream().map(ExerciseResource::from).toList(),
                exercise.getEquipment().stream().map(EquipmentResource::from).toList(),
                exercise.getMainMuscleGroupsIds(), exercise.getSynergisticMuscleGroupsIds());
    }
}

File: src/server/src/main/kotlin/com/project/application/models/exercise/ExerciseDetailsResource.kt
package com.project.application.models.exercise

import com.project.domain.workout.Equipment
import com.project.domain.workout.Exercise
import com.project.domain.workout.MuscleGroup
import io.micronaut.serde.annotation.Serdeable
import java.util.UUID

@Serdeable
data class ExerciseDetailsResource(val id: UUID, val name: String?, val gifUrl: String?, val translatedName: String?,
                              val description: String?, val muscleGroups: List<MuscleGroupDetailsResource>,
                              val variations: List<ExerciseResource>, val equipment: List<EquipmentResource>,
                              val mainMuscleGroupsIds: List<String>, val synergisticMuscleGroupsIds: List<String>) {

    companion object {
        fun from(exercise: Exercise): ExerciseDetailsResource {
            return ExerciseDetailsResource(exercise.id, exercise.name,
                    exercise.gifUrl, exercise.translatedName, exercise.description,
                    exercise.muscleGroups.map { muscleGroup: MuscleGroup -> MuscleGroupDetailsResource.from(muscleGroup) },
                    exercise.variations.map { ex: Exercise -> ExerciseResource.from(ex) },
                    exercise.equipment.map { equipment: Equipment -> EquipmentResource.from(equipment) },
                    exercise.mainMuscleGroupsIds, exercise.synergisticMuscleGroupsIds)
        }
    }
}
